src/server/cached-data.ts

```diff
@@ -0,0 +1,75 @@
+const DEFAULT_TTL_MS = 30000
+
+export type CachedDataOptions = {
+    ttlMs?: number
+}
+
+type Resolver = { resolve: () => void; reject: (error: Error) => void }
+
+export abstract class CachedData<T> {
```

Comment:
We can't be sure, that during highload-induced slowdown `loadActual` will return relevant data in all possible use-cases of CachedData. So I ask you to consider binding ttl to the start of refresh instead of end of refresh.

Please, consider the following implementation:
```typescript
import { Deferred } from "./utils"

const DEFAULT_TTL_MS = 30000

const enum CachedDataExpirationMode {
    ByUpdateStart,
    ByUpdateEnd,
}

export type CachedDataOptions = {
    ttlMs?: number
    expirationMode?: CachedDataExpirationMode
}

export type CachedDataItem<T> = {
    data: Promise<T>
    update_start_time: number
    update_end_time: number | undefined
}

export abstract class CachedData<T> {
    private readonly ttlMs: number
    private readonly expirationMode: CachedDataExpirationMode
    private lastUpdate?: CachedDataItem<T>

    protected constructor(options?: CachedDataOptions) {
        this.ttlMs = options?.ttlMs ?? DEFAULT_TTL_MS
        this.expirationMode =
            options?.expirationMode ?? CachedDataExpirationMode.ByUpdateStart
    }

    abstract loadActual(): Promise<T>

    async get(): Promise<T> {
        const time = Date.now()
        if (this.lastUpdate && !this.is_expired(this.lastUpdate, time)) {
            return this.lastUpdate.data
        }
        return this.refresh()
    }

    async refresh(): Promise<T> {
        const dataWaiter = new Deferred<T>()
        const cachedItem = {
            update_start_time: Date.now(),
            data: dataWaiter.promise,
        } as CachedDataItem<T>

        this.lastUpdate = cachedItem
        try {
            const data = await this.loadActual()
            dataWaiter.resolve(data)
            cachedItem.update_end_time = Date.now()
            return data
        } catch (error) {
            dataWaiter.reject(error)
            cachedItem.update_end_time = Date.now()
            throw error
        }
    }

    private is_expired(dataItem: CachedDataItem<T>, time: number): boolean {
        switch (this.expirationMode) {
            case CachedDataExpirationMode.ByUpdateStart:
                return dataItem.update_start_time + this.ttlMs > time
            case CachedDataExpirationMode.ByUpdateEnd:
                return (
                    !!dataItem.update_end_time &&
                    dataItem.update_end_time + this.ttlMs > time
                )
            default:
                throw Error(`Unknown expiration mode: ${this.expirationMode}`)
        }
    }
}
```